### PR TITLE
Tokenizer's APIs Update

### PR DESCRIPTION
**Pull request by tarekgh**
Updating the Tokenizer's APIs:
- Simplifying the APIs by merging the `Model` abstraction into the `Tokenizer` abstracted class.
- Adding overloads to work with spans.
- doing some clean up optimization and adding more tests.

### CODE REVIEW

**Comment by tarekgh** [comment]
@michaelgsharp please have a look as I have changed the tokenizer APIs and updated the TorchSharp code to work with the new APIs. Thanks!

**Review by ericstj — COMMENTED** [review]
I only put doc comment feedback in the first tokenizer, but it applies to all.  Just a couple other questions.

File: src/Microsoft.ML.Tokenizers/Model/BPE.cs [hunk 1]
```diff
@@ -166,47 +174,326 @@ private Bpe(Stream vocabStream, Stream? mergesStream, string? unknownToken, stri
...
+        /// <param name="tokenCount">The token count can be generated which should be smaller than the maximum token count.</param>
+        /// <param name="considerPreTokenization">Indicate whether to consider pre-tokenization before tokenization.</param>
+        /// <param name="considerNormalization">Indicate whether to consider normalization before tokenization.</param>
+        /// <returns>
+        /// The start index of the maximum encoding capacity within the processed text without surpassing the token limit.
+        /// It represents the index at the first character to be included. In cases where no tokens fit, the result will be length of the text or the <paramref name="normalizedString"/> if normalization is enabled;
+        /// conversely, if all tokens fit, the result will be 0.
+        /// </returns>
+        /// <remarks>
+        /// If the whole text can be encoded within the token limit, the returned index will be 0.
```
[the remark this answers is not in the document]
Comment (tarekgh) [reply src/Microsoft.ML.Tokenizers/Model/BPE.cs:435]: It is redundant. I forgot to delete it after I added the text to the return element. I'll delete the remarks. Thanks!

File: src/Microsoft.ML.Tokenizers/Model/Tiktoken.cs [hunk 2]
```diff
@@ -677,7 +956,7 @@ internal enum ModelEncoding
                                                                 // chat
                                                                 ( "gpt-4-", ModelEncoding.Cl100kBase),          // e.g., gpt-4-0314, etc., plus gpt-4-32k
                                                                 ( "gpt-3.5-turbo-", ModelEncoding.Cl100kBase),  // e.g, gpt-3.5-turbo-0301, -0401, etc.
```
[the remark this answers is not in the document]
Comment (tarekgh) [reply src/Microsoft.ML.Tokenizers/Model/Tiktoken.cs:958]: You are right. I'll remove `turbo-` part from the prefix.

File: src/Microsoft.ML.Tokenizers/Normalizer/Normalizer.cs [hunk 3]
```diff
@@ -17,5 +17,12 @@ public abstract class Normalizer
         /// <param name="original">The original string to normalize.</param>
         /// <returns>The normalized string.</returns>
         public abstract string Normalize(string original);
+
+        /// <summary>
+        /// Process the original string to modify it and obtain a normalized string.
+        /// </summary>
+        /// <param name="original">The original string to normalize.</param>
+        /// <returns>The normalized string.</returns>
+        public virtual string Normalize(ReadOnlySpan<char> original) => Normalize(original.ToString());
```
[the remark this answers is not in the document]
Comment (tarekgh) [reply src/Microsoft.ML.Tokenizers/Normalizer/Normalizer.cs]: I made it abstract. Users will have the choice to do `Normalize(original.ToString());` if they want.

File: src/Microsoft.ML.Tokenizers/PreTokenizer/PreTokenizer.cs [hunk 4]
```diff
@@ -76,6 +76,13 @@ public abstract class PreTokenizer
         /// <returns>The list of the splits containing the tokens and the token's offsets to the original string.</returns>
         public abstract IEnumerable<Split> PreTokenize(string text);
 
+        /// <summary>
+        /// Splits the given character span in multiple spans at the word boundary, keeping track of the offsets of said substrings from the original span.
+        /// </summary>
+        /// <param name="text">The character span to split into tokens.</param>
+        /// <returns>The list of the splits containing the tokens and the token's offsets to the original character span.</returns>
+        public virtual IEnumerable<Split> PreTokenize(ReadOnlySpan<char> text) => PreTokenize(text.ToString());
```
[the remark this answers is not in the document]
Comment (tarekgh) [reply src/Microsoft.ML.Tokenizers/PreTokenizer/PreTokenizer.cs]: I made it abstract.

**Comment by tarekgh** [comment]
I addressed the feedback and did some more optimization too.

**Comment by tarekgh** [comment]
@ericstj @michaelgsharp please let me know if you have any more feedback or we are good to go. Thanks!

File: src/Microsoft.ML.Tokenizers/Normalizer/Normalizer.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (ericstj) [reply src/Microsoft.ML.Tokenizers/Normalizer/Normalizer.cs]: I think that's better.  That way they don't skip it and introduce an extra allocation that can be avoided.

File: src/Microsoft.ML.Tokenizers/Normalizer/UpperCaseNormalizer.cs [hunk 5]
```diff
@@ -22,5 +24,26 @@ public sealed class UpperCaseNormalizer : Normalizer
...
+                return string.Empty;
+            }
+
+            char[] arrayPoolArray = ArrayPool<char>.Shared.Rent(original.Length);
+
+            int length = original.ToUpperInvariant(arrayPoolArray);
+            Debug.Assert(length == original.Length);
+
+            ArrayPool<char>.Shared.Return(arrayPoolArray);
+            return new string(arrayPoolArray, 0, length);
```
Comment (ericstj) [inline src/Microsoft.ML.Tokenizers/Normalizer/UpperCaseNormalizer.cs:46]: Is it OK to use this string after returning it to the pool?  Couldn't someone else get it in between the return and the new string copying its value?

File: test/Microsoft.ML.Tokenizers.Tests/NormalizerTests.cs [hunk 6]
```diff
@@ -129,6 +147,16 @@ public override string Normalize(string original)
...
             }
+
+            public override string Normalize(ReadOnlySpan<char> original)
+            {
+                if (original.IsEmpty)
+                {
+                    return string.Empty;
+                }
+
+                return original.ToString().Normalize(_normalizationForm);
```
Comment (ericstj) [inline test/Microsoft.ML.Tokenizers.Tests/NormalizerTests.cs:158]: Seems like we could `ReadOnlySpan<Char>.Normailze` if we wanted to save an allocation here.  Do you think that's common?

**Review by ericstj — APPROVED** [review]
Looks good if you fix the issue in [UpperCaseNormalizer.cs](https://github.com/dotnet/machinelearning/pull/7128/files/82681543c96d4085730b20e4697df600e6327087..3874f4423417794e6c8d10a86cc46cc596afed42#diff-c86c4a735fee1e554327e7eac4eac622e47c7aa74ebbbc4463433f15f525e63f)

File: src/Microsoft.ML.Tokenizers/Normalizer/UpperCaseNormalizer.cs (hunk 5 above)
Comment (tarekgh) [reply src/Microsoft.ML.Tokenizers/Normalizer/UpperCaseNormalizer.cs:46]: Ops. this was not intentional. Thanks for the catch. We have the same issue in Lowercase one too :-(

File: test/Microsoft.ML.Tokenizers.Tests/NormalizerTests.cs (hunk 6 above)
Comment (tarekgh) [reply test/Microsoft.ML.Tokenizers.Tests/NormalizerTests.cs:158]: Yes this is a test but I believe `ReadOnlySpan<Char>.Normailze` not exist too :-)

**Pull request merged**